source/common/upstream/cluster_manager_impl.cc

```diff
@@ -343,6 +353,86 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   }
 }
 
+bool ClusterManagerImpl::scheduleUpdate(const Cluster& cluster, uint32_t priority,
+                                        const HostVector& hosts_added,
+                                        const HostVector& hosts_removed) {
+  const auto& update_merge_window = cluster.info()->lbConfig().update_merge_window();
+  const auto timeout = DurationUtil::durationToMilliseconds(update_merge_window);
+  PendingUpdatesByPriorityMapPtr updates_by_prio;
+  PendingUpdatesPtr updates;
+
+  // Find pending updates for this cluster.
+  const auto updates_by_prio_it = updates_map_.find(cluster.info()->name());
+  if (updates_by_prio_it != updates_map_.end()) {
+    updates_by_prio = updates_by_prio_it->second;
+  } else {
+    updates_by_prio = std::make_shared<PendingUpdatesByPriorityMap>();
+    updates_map_[cluster.info()->name()] = updates_by_prio;
+  }
+
+  // Find pending updates for this priority.
+  const auto updates_it = updates_by_prio->find(priority);
+  if (updates_it != updates_by_prio->end()) {
+    updates = updates_it->second;
+  } else {
+    updates = std::make_shared<PendingUpdates>();
+    (*updates_by_prio)[priority] = updates;
+  }
+
+  // Has an update_merge_window gone by since the last update? If so, don't schedule
+  // the update so it can be applied immediately as long as there's no active timer
+  // (to avoid changing the order of updates).
+  const auto delta = std::chrono::steady_clock::now() - updates->last_updated;
+  const uint64_t delta_ms = std::chrono::duration_cast<std::chrono::milliseconds>(delta).count();
+  if (delta_ms > timeout && updates->timer == nullptr) {
+    updates->last_updated = std::chrono::steady_clock::now();
+    return false;
+  }
+
+  // Record the updates that should be applied when the timer fires.
+  // Handle the case when a pair of add/remove cancels out.
+  for (const auto& host : hosts_added) {
+    if (updates->removed.count(host) == 1) {
+      updates->removed.erase(host);
+    } else {
+      updates->added.insert(host);
```

Comment:
Do we want to take the latest value for host, since it might have a new health status etc? As this works today, if we do add(unhealthy), add(healthy), only the unhealthy status for the host is set, since insert just returns false the second time.

Again, as with add, remove, add, I think we would benefit from some more tests here.

Reply:
yes, we want the latest value. and yep, will follow-up with comprehensive testing.

Reply:
@htuch note that a health change does not get account as an add/remove:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.cc#L902

That is, if (only) a host's health flipped you'd get called with no new added/removed hosts in which case we'd still do the right thing. Which is, calling `postThreadLocalClusterUpdate()` with empty vectors for `hosts_added` and `hosts_removed`.

Anyway, I'll back all of these claims with tests.